applications/FluidDynamicsApplication/custom_elements/fluid_adjoint_element.cpp

```diff
@@ -0,0 +1,717 @@
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::GetValuesVector(
+    VectorType& rValues,
+    int Step) const
+{
+    if (rValues.size() != TElementLocalSize) {
+        rValues.resize(TElementLocalSize, false);
+    }
+
+    const auto& r_geometry = this->GetGeometry();
+    IndexType local_index = 0;
+    for (IndexType i_node = 0; i_node < TNumNodes; ++i_node) {
+        const auto& r_node = r_geometry[i_node];
+        const auto& r_velocity =
+            r_node.FastGetSolutionStepValue(ADJOINT_FLUID_VECTOR_1, Step);
+        for (IndexType d = 0; d < TDim; ++d) {
+            rValues[local_index++] = r_velocity[d];
+        }
+        rValues[local_index++] =
+            r_node.FastGetSolutionStepValue(ADJOINT_FLUID_SCALAR_1, Step);
+    }
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::GetFirstDerivativesVector(
+    VectorType& rValues,
+    int Step) const
+{
+    if (rValues.size() != TElementLocalSize) {
+        rValues.resize(TElementLocalSize, false);
+    }
+    rValues.clear();
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::GetSecondDerivativesVector(
+    VectorType& rValues,
+    int Step) const
+{
+    if (rValues.size() != TElementLocalSize) {
+        rValues.resize(TElementLocalSize);
+    }
+
+    const auto& r_geometry = this->GetGeometry();
+    IndexType local_index = 0;
+    for (IndexType i_node = 0; i_node < TNumNodes; ++i_node) {
+        const auto& r_acceleration =
+            r_geometry[i_node].FastGetSolutionStepValue(ADJOINT_FLUID_VECTOR_3, Step);
+        for (IndexType d = 0; d < TDim; ++d) {
+            rValues[local_index++] = r_acceleration[d];
+        }
+        rValues[local_index++] = 0.0; // pressure dof
+    }
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::Initialize(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    // If we are restarting, the constitutive law will be already defined
+    if (mpConstitutiveLaw == nullptr) {
+        const Properties& r_properties = this->GetProperties();
+
+        KRATOS_ERROR_IF_NOT(r_properties.Has(CONSTITUTIVE_LAW))
+            << "In initialization of Element " << this->Info()
+            << ": No CONSTITUTIVE_LAW defined for property "
+            << r_properties.Id() << "." << std::endl;
+
+        mpConstitutiveLaw = r_properties[CONSTITUTIVE_LAW]->Clone();
+
+        const GeometryType& r_geometry = this->GetGeometry();
+        const auto& r_shape_functions =
+            r_geometry.ShapeFunctionsValues(GeometryData::GI_GAUSS_1);
+
+        mpConstitutiveLaw->InitializeMaterial(r_properties, r_geometry,
+                                              row(r_shape_functions, 0));
+    }
```

Comment:
If deriving from the `FluidElement` we would avoid this.

Reply:
Because `TAdjointElementData` is different from `TElementData`. I was not able to use same `TElementData` from fluid elements because structure is different.